### PR TITLE
Pawelka/e2etests

### PR DESCRIPTION
**Pull request by moozzyk**
Adding basic end-to-end tests for the ts/js client.

### CODE REVIEW

File: test/Microsoft.AspNetCore.SignalR.Test.Server/Startup.cs [hunk 1]
```diff
@@ -9,7 +9,8 @@ public class Startup
     {
         public void ConfigureServices(IServiceCollection services)
         {
-            services.AddRouting();
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/Startup.cs]: Why did you remove the AddRouting call?

Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/Startup.cs]: It's part of AddSockets and AddSignalR

Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/Startup.cs]: Ah okay. Looks like it was added to AddSockets here https://github.com/aspnet/SignalR/commit/1b59fc6f804e4576e704e6c6f8f9aa90b79ab64d

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+            hubConnection.start(transportName)
+                .then(() => {
+                    hubConnection.invoke('Microsoft.AspNetCore.SignalR.Test.Server.TestHub.ThrowException', errorMessage)
+                        .then(() => {
+                            // exception expected but none thrown
+                            expect(true).toBe(false);
+                        })
+                        .catch(e => {
+                            expect(e.message).toBe(errorMessage);
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: Do we want to (Can we?) check the exception type here?

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        it(`over ${transportName} rethrows an exception from the server`, done => {
+            const errorMessage = "An error occurred.";
+            let hubConnection = new signalR.HubConnection(TESTHUBENDPOINT_URL, 'formatType=json&format=text');
+
+            hubConnection.start(transportName)
+                .then(() => {
+                    hubConnection.invoke('Microsoft.AspNetCore.SignalR.Test.Server.TestHub.ThrowException', errorMessage)
+                        .then(() => {
+                            // exception expected but none thrown
+                            expect(true).toBe(false);
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: nit: Not sure if this was intentional/ has a different meaning but in the test failure case above you say `expect(false).toBe(true) ` and here it is reversed.

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                })
+        });
+
+        it(`over ${transportName} rethrows an exception from the server`, done => {
+            const errorMessage = "An error occurred.";
+            let hubConnection = new signalR.HubConnection(TESTHUBENDPOINT_URL, 'formatType=json&format=text');
+
+            hubConnection.start(transportName)
+                .then(() => {
+                    hubConnection.invoke('Microsoft.AspNetCore.SignalR.Test.Server.TestHub.ThrowException', errorMessage)
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: Does this now just get changed to "ThrowException" since https://github.com/aspnet/SignalR/pull/63 was merged?

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+            client.on("Message", msg => {
+                expect(msg).toBe(message);
+                client.stop();
+                done();
+            });
+
+            client.start(transportName)
+                .then(() => {
+                    client.invoke('Microsoft.AspNetCore.SignalR.Test.Server.TestHub.InvokeWithString', message)
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: Same question here. Should the first param just become "InvokeWithString" ?

File: test/Microsoft.AspNetCore.SignalR.Test.Server/TestHub.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/TestHub.cs]: Copyright and license header here

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/connectionTests.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/connectionTests.js]: Remove leading newline

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js (hunk 4 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: Yes, I had made the changes locally before #63 was merged to actually test #63. I will rebase and update.

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js (hunk 3 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: jasmine does not have `fail` so for the purpose of testing both are equivalent as the mean to fail the test. I think I could just create a fail() function and replace all these.

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js (hunk 2 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: Not really. Exception types from one language don't translate well to exception types in another language and trying to force it would be weird. So we just throw `Error` in javascript. We also don't even send any exception details at the moment (stacktrace and maybe even the exception type might be useful for debugging/tracing)

File: test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js (hunk 5 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/wwwroot/js/hubConnectionTests.js]: As per the comment above - will rebase on dev and update
